designs/CentrallyManagingNuGetPackageVersions-VSManagePackages.md

```diff
@@ -0,0 +1,95 @@
+
+## Scenarios
+
+The following scenarios will be enabled when the proposed design changes will be completed.
+
+1. Installing a package in a project with Package Manager UI or Package Manager Console.
+2. Installing a package from a quick action when editing a source file.
+3. Removing a package from a project with Package Manager UI or Package Manager Console.
+4. Updating the version of a package for all projects using that `Directory.Packages.props` file.
+
+## Requirements
+
+The requirements are defined in the [feature spec](https://github.com/NuGet/Home/wiki/Centrally-managing-NuGet-package-versions).
+
+Note that:
+
+* It's valid for some, not all, projects in the solution to use centrally managed package versions.
+* It's valid for different projects using central package versions to use different `Directory.Packages.props` files.
+* As is valid now, it's valid for some projects to use `packages.config`, while others use `PackageReference`. The `packages.config` projects cannot use centrally managed package versions, and individual `PackageReference` projects may opt in or out of central package version management.
+
+## Solution
+
+### Installing a package in a project
+
+Currently NuGet calls project system APIs to add a PackageReference to the project for the given package id and version. However, when the project uses centrally managed versions, the project file must not add the version on the `PackageReference` element.
+
+There are two categories of solutions.
+
+* API ignorant of centrally managed package versions
+
+Project system APIs would not be changed, but the implementation would need to be aware of how to detect if the feature is enabled, and know when to add a version to the `PackageReference` and when not to, and when to modify the `Directory.Packages.props` and when not not.
+
+There would also be a potential issue when a single user gesture causes a package to be installed into multiple projects, particularly when the `Directory.Packages.props` file does not already contain a `PackageVersion` for the package. If NuGet calls APIs to modify all the projects in parallel, there might be a race condition with trying to update `Directory.Packages.props`.
+
+* API aware of centrally managed package versions
+
+Presumably this would mean there's a separate API to add a `PackageVersion` to the `Directory.Packages.props` file. But since this file is shared across multiple projects, would the API be independent from the projects, or should NuGet pick one project and call the API on that project?
+
+Should new APIs be added to the project systems considering the version attribute must not be written? Or NuGet could call the API with a null version. Or NuGet passes the version and the project system must know if the project uses centrally managed package versions or not.
+
+### Removing a package from a project
```

Comment:
Another section I think it would be regarding the View of versions in PM UI . Is there anything needed from the SDK and / or previous SDK style projects to display the correct CPVM versions?

Reply:
If NuGet has enough information to restore, then it has enough information to display in PMUI. What that really means is that for SDK style projects, we get the PackageVersion items from nomination, and PMUI needs to get it from our own project cache (using our existing project type abstractions). For non-SDK style projects, NuGet will call the new APIs being added for restore, and will do so every time it needs to populate the PMUI list.